### PR TITLE
Issue #1674

### PR DESCRIPTION
**Pull request by mikeljordan**
This PR adds unit-invariance tests for boundary-condition value methods, to verify their SI units using porepy.Units rescaling. Each test applies non-unitary internal unit systems, solves a small probe model, recovers the relevant observable to SI units, and checks that the recovered result is invariant under the rescaling. A mismatch between the declared and true SI dimensions of the BC produces a clean monomial discrepancy in the failure ratio, which makes the diagnostic actionable: the empirical exponent identifies the correct unit.

The covered BC methods are:
- `bc_values_darcy_flux`
- `bc_values_fluid_flux`
- `bc_values_stress`
- `bc_values_fourier_flux`
- `bc_values_enthalpy_flux`
- `bc_values_component_flux`

Docstrings of the affected methods, along with examples and tutorials referencing them, have been updated to reflect the verified SI unit definitions.
Co-authored-by: @RonghaoCui 

## Types of changes

What types of changes does this PR introduce to PorePy?
_Put an `x` in the boxes that apply._

- [ ] Minor change (e.g., dependency bumps, broken links).
- [x] Bugfix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [x] Testing (contribution related to testing of existing or new functionality).
- [x] Documentation (contribution related to adding, improving, or fixing documentation).
- [ ] Maintenance (e.g., improve logic and performance, remove obsolete code).
- [ ] Other:

## Checklist

_Put an `x` in the boxes that apply or explain briefly why the box is not relevant._

- [x] The documentation is up-to-date.
- [x] Static typing is included in the update.
- [x] This PR does not duplicate existing functionality.
- [x] The update is covered by the test suite (including tests added in the PR).
- [ ] If new skipped tests have been introduced in this PR, `pytest` was run with the `--run-skipped` flag.

### CODE REVIEW

File: src/porepy/applications/test_utils/well_models.py
Comment (IvarStefansson) [inline src/porepy/applications/test_utils/well_models.py]: Please update docstring.

File: tests/models/test_boundary_condition.py [hunk 1]
```diff
@@ -10,6 +15,9 @@
     SquareDomainOrthogonalFractures,
 )
 from porepy.applications.test_utils.models import MassBalance as MassBalance_
+
+# from tests.functional.setups.linear_tracer import TracerFlowModel_1p
```
Comment (IvarStefansson) [inline tests/models/test_boundary_condition.py]: Should we remove this line?

File: tests/models/test_boundary_condition.py [hunk 2]
```diff
@@ -451,3 +459,483 @@ class LocalReferenceModel(reference_bc_class, reference_model_class):
...
+    declared_bc_unit: str
+    extraction_subdomain_dim: int = 2
+
+
+#: Unit scalings probing each axis of the BC's SI dimensions.
+_DEFAULT_UNIT_SCALINGS: list[pp.Units] = [
+    pp.Units(m=10.0),
+    pp.Units(kg=0.01),
+    pp.Units(m=100.0),
+    pp.Units(m=10.0, kg=0.01, K=10.0),
```
Comment (IvarStefansson) [inline tests/models/test_boundary_condition.py]: I suggest a more orthogonal set of values to avoid spurious cancellation effects. E.g. m=3, kg=7, K=0.23 or whatever.

File: tests/models/test_boundary_condition.py [hunk 3]
```diff
@@ -451,3 +459,483 @@ class LocalReferenceModel(reference_bc_class, reference_model_class):
...
+        atol=1e-10,
+    )
+
+
+@pytest.mark.parametrize(
+    "spec",
+    _BC_UNIT_INVARIANCE_SPECS,
+    ids=_bc_spec_label,
+)
+def test_bc_values_baseline_well_posed(spec: _BCUnitInvarianceSpec) -> None:
```
Comment (IvarStefansson) [inline tests/models/test_boundary_condition.py]: This test should run on exactly the same models as the other one you introduced, right? If that is the case, we should avoid rerunning the simulations to save test time. I have two suggestions for how we could achieve that (but am open to other approaches): 1 insert the check into the other test or 2 convert the construction (including the running of the simulations) into a parametrized fixture?

File: tests/models/test_boundary_condition.py [hunk 4]
```diff
@@ -451,3 +459,483 @@ class LocalReferenceModel(reference_bc_class, reference_model_class):
...
+    "spec",
+    _BC_UNIT_INVARIANCE_SPECS,
+    ids=_bc_spec_label,
+)
+@pytest.mark.parametrize(
+    "units",
+    _DEFAULT_UNIT_SCALINGS,
+    ids=_unit_scaling_label,
+)
+def test_bc_values_unit_invariance(
```
Comment (IvarStefansson) [inline tests/models/test_boundary_condition.py:912]: Do you agree we could mark this test as slow (weekly run instead of each PR)?

**Review by IvarStefansson — COMMENTED** [review]
Nice work. Some relatively minor comments only. 
One is included here to circumvent notebook reviewing, which is a pain: Please add units of flux BCs in docstrings in boundary_conditions.ipynb.

File: tests/models/test_boundary_condition.py (hunk 2 above)
Comment (mikeljordan) [reply tests/models/test_boundary_condition.py]: Yes, this won't be a problem, the test should hold for any values of m, kg, K. Will do that now.

File: tests/models/test_boundary_condition.py (hunk 3 above)
Comment (mikeljordan) [reply tests/models/test_boundary_condition.py]: Good point. I’ll fold the check into the unit-invariance assertion, following your first suggestion and remove the redundant test.

File: tests/models/test_boundary_condition.py (hunk 4 above)
Comment (mikeljordan) [reply tests/models/test_boundary_condition.py:912]: Yes, I agree. I’ll add the slow marker to the test.

File: tests/models/test_boundary_condition.py (hunk 1 above)
Comment (RonghaoCui) [reply tests/models/test_boundary_condition.py]: Yes, this line has been removed.

File: src/porepy/applications/test_utils/well_models.py
Comment (RonghaoCui) [reply src/porepy/applications/test_utils/well_models.py]: The docstrings on units of flux BCs in the tutorial have been added. 
The boundary condition docstrings in well_models.py have also been updated. A new review for well_models.py is requested.

File: tests/models/test_boundary_condition.py (hunk 2 above)
Comment (IvarStefansson) [reply tests/models/test_boundary_condition.py]: Thanks. Just to clarify: I was fine with not changing all units for all types. I was only referring to the last case, the one which had non-default values for all three units.  Feel free to resolve again if you want to keep as is or ask if I'm being confusing.

File: tutorials/boundary_conditions.ipynb [hunk 5]
```diff
@@ -370,7 +380,13 @@
                 "\n",
                 "    # Note that now this is bc_values_darcy_flux, not the bc_values_fluid_flux.\n",
                 "    def bc_values_darcy_flux(self, bg: pp.BoundaryGrid) -> np.ndarray:\n",
-                "        \"\"\"Setting the Darcy flux values on the west boundary.\"\"\"\n",
+                "        \"\"\"Setting the mobility-free Darcy flux values on the west boundary.\n",
+                "        \n",
+                "        The values correspond to the integrated quantity of mobility-free Darcy \n",
+                "        flux over the boundary face on Neumann boundary. In two dimensions, the \n",
+                "        SI unit of integrated mobility-free Darcy flux is m^2*Pa.\n",
```
Comment (IvarStefansson) [inline tutorials/boundary_conditions.ipynb]: Great, this will be much appreciated! Optional suggestion: Since this is a tutorial, should we be nice to the user and spell out the 3d case as well?

**Review by IvarStefansson — APPROVED** [review]
Thanks. This is going to be useful for users new and old!

Please have a look at the new comments, then merge.

File: tutorials/boundary_conditions.ipynb (hunk 5 above)
Comment (mikeljordan) [reply tutorials/boundary_conditions.ipynb]: Yes I agree, we will include that
